File: packages/cordova/src/android/cordova/ads/Native.java
package admob.plus.cordova.ads;

import android.content.Context;
import android.graphics.Color;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdLoader;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.nativead.NativeAd;

import java.util.HashMap;
import java.util.Map;

import admob.plus.cordova.ExecuteContext;
import admob.plus.cordova.Generated.Events;

public class Native extends AdBase implements GenericAd {
    public static final Map<String, ViewProvider> providers = new HashMap<String, ViewProvider>();

    private final AdRequest mAdRequest;
    private final ViewProvider viewProvider;
    private AdLoader mLoader;
    private NativeAd mAd;
    private View view;

    public Native(ExecuteContext ctx) {
        super(ctx);

        mAdRequest = ctx.optAdRequest();
        String key = "default";
        viewProvider = providers.get(key);
        if (viewProvider == null) {
            throw new RuntimeException("cannot find viewProvider: " + key);
        }
    }

    @Override
    public void onDestroy() {
        clear();

        super.onDestroy();
    }

    @Override
    public boolean isLoaded() {
        return mLoader != null && !mLoader.isLoading();
    }

    @Override
    public void load(ExecuteContext ctx) {
        clear();

        mLoader = new AdLoader.Builder(getActivity(), adUnitId)
                .forNativeAd(nativeAd -> {
                    mAd = nativeAd;
                })
                .withAdListener(new AdListener() {
                    @Override
                    public void onAdFailedToLoad(LoadAdError adError) {
                        emit(Events.AD_LOAD_FAIL, adError);
                        if (isLoaded()) {
                            ctx.error(adError.toString());
                        }
                    }

                    public void onAdClosed() {
                        emit(Events.AD_DISMISS);
                    }

                    public void onAdOpened() {
                        emit(Events.AD_SHOW);
                    }

                    public void onAdLoaded() {
                        emit(Events.AD_LOAD);
                        if (isLoaded()) {
                            ctx.success();
                        }
                    }

                    public void onAdClicked() {
                        emit(Events.AD_CLICK);
                    }

                    public void onAdImpression() {
                        emit(Events.AD_IMPRESSION);
                    }
                })
                .build();
        mLoader.loadAd(mAdRequest);
    }

    @Override
    public void show(ExecuteContext ctx) {
        if (view == null) {
            view = viewProvider.createView(getActivity(), mAd);
            ViewGroup vg = getWebViewParent();
            vg.addView(view);

            view.setBackgroundColor(Color.parseColor("#ffffff"));
        }

        view.setVisibility(View.VISIBLE);
        view.setX((float) ctx.opts.optDouble("x"));
        view.setY((float) ctx.opts.optDouble("y"));
        ViewGroup.LayoutParams params = view.getLayoutParams();
        params.width = (int) ctx.opts.optDouble("width");
        params.height = (int) ctx.opts.optDouble("height");
        view.setLayoutParams(params);

        viewProvider.didShow(this);

        view.requestLayout();
    }

    @Override
    public void hide(ExecuteContext ctx) {
        if (view != null) {
            view.setVisibility(View.GONE);
        }

        viewProvider.didHide(this);
        ctx.success();
    }

    private void clear() {
        if (mAd != null) {
            mAd.destroy();
            mAd = null;
        }
        mLoader = null;
    }

    public interface ViewProvider {
        @NonNull
        View createView(Context context, NativeAd nativeAd);

        default void didShow(@NonNull Native ad) {
        }

        default void didHide(@NonNull Native ad) {
        }
    }
}

File: packages/cordova/src/android/cordova/ads/Native.kt
package admob.plus.cordova.ads

import admob.plus.cordova.Events
import admob.plus.cordova.ExecuteContext
import admob.plus.core.buildAdRequest
import admob.plus.core.dpToPx
import android.util.Log
import android.view.View
import android.view.ViewGroup
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdLoader
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import java.util.Objects
import java.util.concurrent.ConcurrentHashMap

class Native(ctx: ExecuteContext) : AdBase(ctx) {
    private val mAdRequest: AdRequest = buildAdRequest(initOpts)
    private val viewProvider: ViewProvider
    private var mLoader: AdLoader? = null
    private var mAd: NativeAd? = null
    private var view: View? = null

    init {
        val key = initOpts.optString("view").ifEmpty { VIEW_DEFAULT_KEY }
        viewProvider = providers[key] ?: throw RuntimeException("cannot find viewProvider: $key")
    }

    override fun onDestroy() {
        clear()
        super.onDestroy()
    }

    override val isLoaded: Boolean get() = mLoader != null && !mLoader!!.isLoading

    override fun load(ctx: ExecuteContext) {
        clear()
        mLoader = AdLoader.Builder(plugin.activity, adUnitId)
            .forNativeAd { nativeAd -> mAd = nativeAd }
            .withAdListener(object : AdListener() {
                override fun onAdFailedToLoad(adError: LoadAdError) {
                    emit(Events.AD_LOAD_FAIL, adError)
                    if (isLoaded) {
                        ctx.reject(adError.toString())
                    }
                }

                override fun onAdClosed() {
                    emit(Events.AD_DISMISS)
                }

                override fun onAdOpened() {
                    emit(Events.AD_SHOW)
                }

                override fun onAdLoaded() {
                    emit(Events.AD_LOAD)
                    if (isLoaded) {
                        ctx.resolve()
                    }
                }

                override fun onAdClicked() {
                    emit(Events.AD_CLICK)
                }

                override fun onAdImpression() {
                    emit(Events.AD_IMPRESSION)
                }
            })
            .build().also {
                it.loadAd(mAdRequest)
            }
    }

    override fun show(ctx: ExecuteContext) {
        val ad = mAd ?: return ctx.reject("ad not loaded")
        view = view ?: let {
            val v = viewProvider.createView(ad)
            Objects.requireNonNull<ViewGroup>(plugin.contentView).addView(v)
            v
        }
        view?.let {
            it.visibility = View.VISIBLE
            it.x = dpToPx(initOpts.optDouble("x", 0.0)).toFloat()
            it.y = dpToPx(initOpts.optDouble("y", 0.0)).toFloat()

            val params = it.layoutParams
            params.width = dpToPx(initOpts.optDouble("width", 0.0)).toInt()
            params.height = dpToPx(initOpts.optDouble("height", 0.0)).toInt()
            it.layoutParams = params

            viewProvider.didShow(this)
            it.requestLayout()
        }
        ctx.resolve(true)
    }

    override fun hide(ctx: ExecuteContext) {
        view?.let {
            it.visibility = View.GONE
        }
        viewProvider.didHide(this)
        ctx.resolve()
    }

    private fun clear() {
        mAd?.let {
            it.destroy()
            mAd = null
        }
        view?.let {
            when (it) {
                is NativeAdView -> {
                    it.removeAllViews()
                    it.destroy()
                }
            }
            view = null
        }
        mLoader = null
    }

    interface ViewProvider {
        fun createView(nativeAd: NativeAd): View

        fun didShow(ad: Native) {
            Log.d(TAG, "Show Ad: ${ad.id}")
        }

        fun didHide(ad: Native) {
            Log.d(TAG, "Hide Ad: ${ad.id}")
        }
    }

    companion object {
        private const val TAG = "AdMobPlus.Native"

        const val VIEW_DEFAULT_KEY = "default"
        val providers = ConcurrentHashMap<String, ViewProvider>()
    }
}
